models/inh_poisson_generator.cpp

```diff
@@ -0,0 +1,213 @@
+  {
+    throw BadProperty( "Rate times and values have to be the same size." );
+  }
+
+  // ensure amp times are strictly monotonically increasing
+  if ( rate_times_.empty() == false )
+  {
+    std::vector< double_t >::const_iterator prev = rate_times_.begin();
+    std::vector< double_t >::const_iterator next = prev + 1;
+    for ( ; next != rate_times_.end(); ++next, ++prev )
+    {
+      if ( *prev >= *next )
+      {
+        throw BadProperty( "Rate times must strictly increasing." );
+      }
+    }
+  }
+
+  if ( ut && uv )
+  {
+    b.idx_ = 0; // reset if we got new data
+  }
+}
+
+/* ----------------------------------------------------------------
+ * Default and copy constructor for node
+ * ---------------------------------------------------------------- */
+
+nest::inh_poisson_generator::inh_poisson_generator()
+  : Node()
+  , device_()
+  , P_()
+{
+}
+
+nest::inh_poisson_generator::inh_poisson_generator(
+  const inh_poisson_generator& n )
+  : Node( n )
+  , device_( n.device_ )
+  , P_( n.P_ )
+{
+}
+
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+void
+nest::inh_poisson_generator::init_state_( const Node& proto )
+{
+  const inh_poisson_generator& pr = downcast< inh_poisson_generator >( proto );
+
+  device_.init_state( pr.device_ );
+}
+
+void
+nest::inh_poisson_generator::init_buffers_()
+{
+  device_.init_buffers();
+  B_.idx_ = 0;
+  B_.rate_ = 0;
+}
+
+void
+nest::inh_poisson_generator::calibrate()
+{
+  device_.calibrate();
+  V_.h_ = Time::get_resolution().get_ms();
+}
+
+/* ----------------------------------------------------------------
+ * Update function and event hook
+ * ---------------------------------------------------------------- */
+
+void
+nest::inh_poisson_generator::update( Time const& origin,
+  const long from,
+  const long to )
+{
+  assert(
+    to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( from < to );
+  assert( P_.rate_times_.size() == P_.rate_values_.size() );
+
+  const long t0 = origin.get_steps();
+
+  // random number generator
+  librandom::RngPtr rng = kernel().rng_manager.get_rng( get_thread() );
+
+
+  // Skip any times in the past. Since we must send events proactively,
+  // idx_ must point to times in the future.
+  const long first = t0 + from;
+  while ( B_.idx_ < P_.rate_times_.size()
+    && Time( Time::ms( P_.rate_times_[ B_.idx_ ] ) ).get_steps() <= first )
+  {
+    ++B_.idx_;
+  }
+
+  for ( long offs = from; offs < to; ++offs )
+  {
+    const long curr_time = t0 + offs;
+
+    // Keep the amplitude up-to-date at all times.
+    // We need to change the amplitude one step ahead of time, see comment
+    // on class SimulatingDevice.
+    if ( B_.idx_ < P_.rate_times_.size()
+      && curr_time + 1
+        == Time( Time::ms( P_.rate_times_[ B_.idx_ ] ) ).get_steps() )
```

Comment:
What will happen here if the rate changes twice within a time step? Since you can only increase `B_.idx_` by one for each value of `offs`, you would probably get out of step for the remainder of the simulation.

Note that this is a deep problem: which rate should you use if the rate changes multiple times within an interval. Indeed, the question is also how to handle rate changes inside an interval correctly. This is especially critical to ensure consistent results when simulating the same network with different resolutions. The `sinusoidal_poisson_generator` always uses the rate at the end of the time step, so a consistent solution would be to always use the rate valid at then end of the step for the entire step. This needs to be documented, and you need code here that skips far enough.